File: ReactNativeApp/ios/Modules/ObjectRepresentation/Locations/Instances/PointLocation.swift
//
//  PointLocation.swift
//  jogo
//
//  Created by arham on 05/03/2021.
//

import Foundation

class PointLocation: DetectionLocation {
  
  // Used for landmarks or points detected. Points are represented by x,y,z?.
  
  private var radius: Double
  private var z: Double
  
  // MARK: Lifecycle
  init(classLabel: String,
       centerX: Double,
       centerY: Double,
       frameId: Int,
       confidence: Float,
       status: STATUS,
       radius: Double,
       zValue: Double = 0) {
    self.radius = radius
    self.z = zValue
    super.init(label: classLabel, centerX: centerX, centerY: centerY,
               frameId: frameId, confidence: confidence, status: status)
  }
  
  init(classLabel: String,
       centerX: Double,
       centerY: Double,
       frameId: Int,
       confidence: Float,
       zValue: Double = 0) {
    self.radius = 0.01
    self.z = zValue
    super.init(label: classLabel, centerX: centerX, centerY: centerY, frameId: frameId,
               confidence: confidence, status: STATUS.DETECTED)
  }
  
  required init(from decoder: Decoder) throws {
    fatalError("init(from:) has not been implemented")
  }
  
  public override func updateLocation(centerX: Double,
                                      centerY: Double,
                                      status: DetectionLocation.STATUS) {
    self.setX(value: centerX)
    self.setY(value: centerY)
    self.status = status
  }
  
  // MARK: Properties Get/Set
  public override func getRadius() -> Double {
    return self.radius
  }
  
  public func getZValue() -> Double {
    return self.z
  }
  
  public func setRadius(value: Double) {
    self.radius = value
  }
  
  override func getCenter() -> (x: Double, y: Double) {
    return (x: self.getX(), y: self.getY())
  }
  
  override func getPoint() -> CGPoint {
    return CGPoint(x: self.getX(), y: self.getY())
  }
  
  // MARK: Drawing
  
  public override func draw(canvas: UIView) {
    
  }
  
  public override func drawDebug(canvas: UIView) {
    
  }
}

File: ReactNativeApp/android/app/src/main/java/com/location/PointLocation.java
package com.location;

import android.graphics.Canvas;
import android.graphics.Paint;

import com.utils.Gear;
import com.utils.Paint443;

public class PointLocation extends DetectionLocation {
    private double radius;

    public PointLocation(String classLabel, double centerX, double centerY, double centerZ, int frameID, float confidence, STATUS status, double radius) {
        super(classLabel, centerX, centerY, centerZ, frameID, confidence, status);
        this.radius = radius;
    }

    public PointLocation(String classLabel, double centerX, double centerY, double centerZ, int frameID, float confidence) {
        this(classLabel, centerX, centerY, centerZ, frameID, confidence, STATUS.DETECTED, 0.01);
    }


    public void updateLocation(double centerX, double centerY, STATUS status) {
        this.x = centerX;
        this.y = centerY;
        this.status = status;

    }


    public void drawNames(Canvas canvas, Paint paint) {
        // draw circle for names
        drawPoints(canvas, paint);
        // draw names
        canvas.drawText(label + " " + String.format("%.2f", confidence), (int) ((getX() * canvas.getWidth()) + radius * Math.min(canvas.getWidth(), canvas.getHeight())), (int) ((getY() * canvas.getHeight()) + radius * Math.min(canvas.getWidth(), canvas.getHeight())), new Paint443().red().fillStroke().small().monospace());
    }

    public void drawPoints(Canvas canvas, Paint paint) {
        canvas.drawCircle((float) getX() * canvas.getWidth(),
                (float) getY() * canvas.getHeight(),
                (float) (radius * Math.min(canvas.getWidth(), canvas.getHeight())),
                paint);
    }

    public void drawDebug(Canvas canvas) {
        drawDebug(canvas, new Paint443().transparancy(.95));
    }


    @Override
    public void drawDebug(Canvas canvas, Paint paint) {

        if (Gear.iff("points", true))
            drawPoints(canvas, paint);
        if (Gear.iff("names", false))
            drawNames(canvas, new Paint443().red().narrowStroke().medium());

    }

    public void draw(Canvas canvas) {
        canvas.drawCircle((float) getX() * canvas.getWidth(),
                (float) getY() * canvas.getHeight(),
                (float) (radius * Math.min(canvas.getWidth(), canvas.getHeight())),
                new Paint443().red().narrowStroke().medium());
    }

    public double getRadius() {
        return radius;
    }

    public void setRadius(int radius) {
        this.radius = radius;
    }
}
